ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -5008,10 +5009,27 @@ static private boolean needToCopy(final HiveConf conf, Path srcf, Path destf, Fi
         boolean isOwned = FileUtils.isOwnerOfFileHierarchy(srcFs, srcs, configuredOwner, false);
         if (configuredOwner.equals(runningUser)) {
           // Check if owner has write permission, else it will have to copy
-          if (!(isOwned &&
-              FileUtils.isActionPermittedForFileHierarchy(
-                  srcFs, srcs, configuredOwner, FsAction.WRITE, false))) {
-            return true;
+          UserGroupInformation ugi = Utils.getUGI();
+          String currentUser = ugi.getShortUserName();
+          FileSystem fsAsUser = null;
+          UserGroupInformation proxyUser = null;
+          try {
+            if (configuredOwner != null && !configuredOwner.equals(currentUser)) {
+              proxyUser = UserGroupInformation.createProxyUser(configuredOwner, UserGroupInformation.getLoginUser());
+              fsAsUser = FileUtils.getFsAsUser(srcFs, proxyUser);
+            }
+            if (!(isOwned && FileUtils.isActionPermittedForFileHierarchy(srcFs, srcs, configuredOwner, FsAction.WRITE,
+                false, fsAsUser))) {
+              if (proxyUser != null) {
+                FileSystem.closeAllForUGI(proxyUser);
```

Comment:
This is redundant ..finally block will close the filesystem

Reply:
Agree